miso-service/src/main/java/uk/ac/bbsrc/tgac/miso/service/impl/DefaultExperimentService.java

```diff
@@ -50,11 +57,60 @@ public void deleteExperiment(Experiment experiment) throws IOException {
   }
 
   @Override
+  public Experiment getExperimentById(long experimentId) throws IOException {
+    Experiment o = experimentStore.get(experimentId);
+    authorizationManager.throwIfNotReadable(o);
+    return o;
+  }
+
+  @Override
+  public Map<String, Integer> getExperimentColumnSizes() throws IOException {
+    return experimentStore.getExperimentColumnSizes();
+  }
+
+  public NamingScheme getNamingScheme() {
+    return namingScheme;
+  }
+
+  @Override
+  public Collection<Experiment> listAllExperiments() throws IOException {
+    return authorizationManager.filterUnreadable(experimentStore.listAll());
+  }
+
+  @Override
+  public Collection<Experiment> listAllExperimentsBySearch(String query) throws IOException {
+    return authorizationManager.filterUnreadable(experimentStore.listBySearch(query));
+  }
+
+  @Override
+  public Collection<Experiment> listAllExperimentsByStudyId(long studyId) throws IOException {
+    return authorizationManager.filterUnreadable(studyStore.get(studyId).getExperiments());
+  }
+
+  @Override
+  public Collection<Experiment> listAllExperimentsWithLimit(long limit) throws IOException {
+    return authorizationManager.filterUnreadable(experimentStore.listAllWithLimit(limit));
+  }
+
+  @Override
   public long saveExperiment(Experiment experiment) throws IOException {
+    try {
     authorizationManager.throwIfNotWritable(experiment);
     if (experiment.getId() == ExperimentImpl.UNSAVED_ID) {
+        String name = namingScheme.generateNameFor(experiment);
```

Comment:
ID is used in names, so entity should be saved first like in HibernateSampleDao; otherwise, every name will be EXP0